TechnicalDocuments/SnowplowHowTo.md

```diff
@@ -0,0 +1,163 @@
+
+```swift
+extension ActionEvents.AchievementUnlocked: UnstructuredEventConvertable {
+    var specificEventDictionary: [String: Any] {
+        switch self {
+        case let .honorReward(_, honor):
+            return ["honor": honor]
+        case let .moneyReward(_, gold):
+            return ["money": gold]
+        }
+    }
+
+    var genericEventDictionary: [String: Any] {
+        let action: String
+
+        switch self {
+        case let .honorReward(name, _):
+            action = name
+        case let .moneyReward(name, _):
+            action = name
+        }
+        return [
+            "category": "achievement",
+            "action": action
+        ]
+    }
+
+    var schema: String { return "iglu:com.thegame/achievement/jsonschema/1-0-0" }
+}
+```
+
+The generic event dictionary may also contain `label`, `property` and `value` where `label` and `property` are strings and `value` is a double. The `specificEventDictionary` dictionary must be consistent with the JSON defined by `schema`. Finally, the event type needs to be added to the dispatch table.
+
+```swift
+fileprivate func makeDispatcher() -> AnalyticsEventDispatcher {
+    let dispatcher = AnalyticsEventDispatcher()
+        .forward(ActionEvents.AchievementUnlocked.self,to: SnowplowTracker.track)
+}
+
+extension SnowplowTracker {
+    func track(_ event: ActionEvents.AchievementUnlocked) {
+        realTracker.trackUnstructuredEvent(event)
+    }
+}
+```
+
+## Unit tests for Snowplow events
+
+There should be unit tests that verify that the unstructured event content is correct.
+
+```swift
+class AchievementUnlockedEventTests: XCTestCase {
+    var mockEnvironment = MockSnowplowTrackingEnvironment()
+    var mock: MockSnowplowTracking { return mockEnvironment.mockSnowplowTracking }
+
+    override func tearDown() {
+        mockEnvironment.reset()
+    }
+
+    func testConvertHonorAwarded() {
+        var correctCall = false
+        mock.trackUnstructuredEventAssertion = { unstructered in
+            guard let json = unstructered.json else { return }
+
+            expect(json.contains("honor")).to(beTrue())
+
+            correctCall = true
+        }
+
+        mockEnvironment.analyticesService.track(ActionEvents.AchievementUnlocked.honorReward(name: "Unit Test Written", honor: 50))
+        expect(correctCall).to(beTrue())
+    }
+```
+
+How detailed the content verification needs to be is a judgement call.
+
+## UI Tests for Snowplow Events.
+
+That the events are fired with the correct generic content should be verified in a UI tests. For performance reasons it is preferable to add analytics verification to an existing UI test.
+
+```swift
+class AchievementUnlockedTests: BaseFeature {
+
+    override func setUp() {
+        super.setUp()
+        self.enableMockSnowplowServer()
+    }
+
+    override func tearDown() {
+        XCTAssert(snowplowServer.verifySnowplowEvents(), "Expected analytics events were not posted")
```

Comment:
I'd like to see as a part of this proposal a more generic way of validating analytics events. For example this can be done in `tearDown` of `BaseFeature`.

Reply:
There are pros and cons with either approach. Maybe we can look at this together when you start work on adding events for `Appointments`? It would be good to leverage your experience with the mock server.